Test on GitHub Actions

hugovk opened the pull request:
Travis CI has a new pricing model which places limits on open source.

* https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing
* https://www.jeffgeerling.com/blog/2020/travis-cis-new-pricing-plan-threw-wrench-my-open-source-works

Many projects are moving to GitHub Actions instead, which provides testing on Linux, macOS and Windows, with 20 parallel jobs (compared to 5 on Travis and 1 on AppVeyor).

This PR adds testing for modern Python 3.6-3.9 and PyPy3 on the three operating systems. I didn't bother with EOL Pythons here.

Here's an example run:

* https://github.com/hugovk/pyflakes/actions/runs/399197614

Note PyPy3/Windows is excluded for the time being, we're waiting for GHA to include the latest PyPy3 release:

* https://github.com/actions/setup-python/issues/163

I also didn't remove Travis CI or AppVeyor here, but I would suggest doing so.

asottile commented:
we should either drop python 2 or add python 2 to this PR and delete the other providers

hugovk commented:
I've got another branch ready to drop Python 2.7, 3.4 and 3.5.

Shall I merge it in here, or add 2.7 to this, or wait for this to be merged then open a new PR?

hugovk commented:
> ... or add python 2 to this PR and delete the other providers

Updated to do this, and also added 3.5. (Python 3.4 isn't available on GHA.)

I think it'll be better to have separate PRs, as the other one is quite big. 

It also means the GHA config would already merged into master, which is needed to enable it for this repo.

asottile commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,60 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["2.7", "3.5", "3.6", "3.7", "3.8", "3.9", "pypy3"]
```
this would need pypy as well

asottile commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,60 @@
+      fail-fast: false
+      matrix:
+        python-version: ["2.7", "3.5", "3.6", "3.7", "3.8", "3.9", "pypy3"]
+        os: [ubuntu-latest, macos-latest, windows-latest]
```
I don't think macos grants us any additional coverage and the macos builders are pretty slow and in high demand -- after adding pypy we're looking at 24 builders which seems excessive to me

we can probably get by with:
- test all of them on linux
- test py2 + minimum py3 + maximum py3 + pypy + pypy3 on windows

hugovk replied:
Thanks for the review!

Note that pypy is broken on GHA until they include the latest pypy release (see https://github.com/actions/setup-python/issues/163), so pypy2/3 are excluded for now. I've included a comment explaining why.

asottile commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,60 @@
+            ${{ matrix.os }}-${{ matrix.python-version }}-v1-${{
+            hashFiles('**/setup.py') }}
+          restore-keys: |
+            ${{ matrix.os }}-${{ matrix.python-version }}-v1-
```
this isn't needed or helpful here -- pyflakes has no dependencies and caching based on the setup.py doesn't gain us anything

asottile commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,60 @@
+        run: |
+          python -m pip install -U pip
+          python -m pip install -U wheel
+          python -m pip install -U tox
```
at the very least, the last two lines can be combined -- though I don't think it's useful to upgrade pip or wheel anyway so this can probabl just be `pip install wheel`

asottile commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,60 @@
+
+      - name: Tox tests
+        shell: bash
+        run: |
+          tox -e py
```
```suggestion
        run: tox -e py
```

hugovk commented:
Ha, AppVeyor fails with no `.appveyor.yml`. It'll need disabling, or at least this option ticking at https://ci.appveyor.com/project/sigmavirus24/pyflakes/settings:

![image](https://user-images.githubusercontent.com/1324225/101086717-f0329200-35b9-11eb-81fa-56e3fa462759.png)

asottile commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,60 @@
+          tox -e py
+
+      - name: Upload coverage
+        uses: codecov/codecov-action@v1
```
we're not using `codecov`

asottile commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,60 @@
+on: [push, pull_request]
+
+env:
+  FORCE_COLOR: 1
```
does anything use this?  this isn't a standard environment variable I'm familiar with

hugovk replied:
At least pytest does, I normally just include it as a default. But it's not being passed via tox and the test output is monochrome, so I'll just remove it.

asottile commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,37 @@
+        run: python -m pip install -U tox
+
+      - name: Tox tests
+        shell: bash
```
this line can be removed too

asottile reviewed (approved):
![](https://i.fluffy.cc/b6wlBbdVkX83CHpLWxHZw74WJxHVQc9S.gif)
